### PR TITLE
logging: Add support for additional logger filters other than hostname

### LINKED ISSUE #5978: [RFE] filter/split access logs by fields other than hostname
Hi!

Currently, I can create an access log like so:

```
    log {
        output file /opt/servers/server1/logs/caddy.access.log
    }
```

However, with many health checks coming in, I would redirect all known User-Agents for health checks to its own log file:

```
    log access_log {
        filter not @health-user-agents && not @health-url-paths
        output file /opt/servers/server1/logs/caddy.access.log
    }
    log health_check_log {
        filter @health-user-agents || @health-url-paths
        output file /opt/servers/server1/logs/caddy.access.health.log
    }
```

Of course this is pseudo syntax, but I think you get the idea.

For reference, in Apache httpd I can write something like this:

```
SetEnvIf Request_URI ".*\/health-check$" HealthPattern
# just add more here, they are or'd

CustomLog     /opt/servers/server1/logs/caddy.access.health.log    combined    env=HealthPattern
# and for access.log:  env!HealthPattern
```

### PR DESCRIPTION
**Pull request by armadi1809**
closes #5978 

@francislavoie This took some time but here is my initial implementation of this. Would like to get your feedback on this and your thoughts on what other tests should go in here? Thanks.

### CODE REVIEW

**Comment by francislavoie** [comment]
Like I said in the issue, I'm not totally convinced this is the right way to do this. But I can't really think of anything better.

I think `no_hostname` is better than `no_host_name`, because `hostname` is typically seen as a single word in the context web servers (see https://en.wikipedia.org/wiki/Hostname for example)

I think there should be a `d.NextArg()` when parsing that option to reject the config if an extra argument is passed, since it takes none right now.

I'm not sure if that's the best name though, I'd like @mholt's opinion.

File: modules/caddyhttp/logging.go
Comment (armadi1809) [reply modules/caddyhttp/logging.go:61]: Yeah I agree that appending would go against the purpose of this feature, so overriding seems reasonable. We can wait until your PR gets merged and discuss how we can make `access_logger_name` take multiple values so that it fits in the multiple logger feature addition in your PR.

**Comment by mholt** [comment]
@francislavoie @armadi1809  Let me know if you want to discuss anything to wrap up this PR. :+1:

**Comment by armadi1809** [comment]
If I remember correctly this needs to wait for #6088 to be merged as it will require some changes after that's done right? @francislavoie

**Comment by francislavoie** [comment]
I'm merging it now (once CI passes) 👍 you can rebase your branch and make the necessary adjustments.

File: modules/caddyhttp/logging.go
Comment (armadi1809) [reply modules/caddyhttp/logging.go]: @francislavoie I went the naive route by making `access_logger_names` take multiple values seperated by a `,`. Do you recall what was your idea here and if we should rather implement it? I was thinking about tweaking the logic of the `UnmarshalCaddyFile` function of the `VarsMiddleware` but I wanted to get your opinion first.

**Comment by francislavoie** [comment]
Needs another rebase, there's a conflict with another recent change of mine. Sorry!

**Comment by armadi1809** [comment]
@francislavoie, rebase done. Any thoughts on the open discussion above?

File: caddyconfig/httpcaddyfile/builtins.go
Comment (francislavoie) [inline caddyconfig/httpcaddyfile/builtins.go:918]: I guess this variable name is misleading since we chose `no_hostname` as the config option

File: modules/caddyhttp/logging.go
Comment (francislavoie) [inline modules/caddyhttp/logging.go:89]: Normally we use `camelCase` for variable names. Anyway, we can shorten them because what it is is obvious from the source it's pulled from.

```suggestion
	if names := GetVar(req.Context(), AccessLoggerNameVarKey); names != nil {
		namesSlice := strings.Split(names.(string), ",")
		for _, loggerName := range namesSlice {
			loggers = append(loggers, logger.Named(loggerName))
		}
		return loggers
	}
```

File: caddytest/integration/caddyfile_adapt/log_filter_with_header.txt
Comment (francislavoie) [inline caddytest/integration/caddyfile_adapt/log_filter_with_header.txt:15]: Oof, yeah this is weird. I don't think using a comma-separated value is ideal. Nowhere else in Caddy do we have this pattern.

@mholt I had an idea that we could do string slices for `vars` if we have the config like this:

```
vars access_logger_names [ health_check_log general_log ]
```

Basically if the first and last token are `[` `]` then we can take the list of args in between as a string slice type value.

I think we should make adjustments to `vars` to make it possible to have slices.

Comment (mholt) [reply caddytest/integration/caddyfile_adapt/log_filter_with_header.txt:15]: Hmm; I see what you mean. But it looks really bash-y. I wonder if we could support `["health_check_log", "general_log"]`...

I actually like how clean and simple the CSV is though. :man_shrugging:

Comment (francislavoie) [reply caddytest/integration/caddyfile_adapt/log_filter_with_header.txt:15]: Now I'm leaning towards adding a new directive `log_name <names...>` which would handle this case for us. That way users don't need to discover the `access_logger_names` var which is kinda a hack, they could just use `log_name` which would be documented like any other directive. This is similar to how `root` works, it's just a directive shortcut for a `vars` handler that sets a variable. I think it would need to be a real handler in JSON though because the `vars` handler still doesn't support an array of values.

Another long-term thing we could consider (out of scope of this PR but just thinking out loud), we could support a syntax like `vars +var_name value`, where `+` means "assume the var is a slice, and append a value to it". Could also add a `-` prefix to clear variable altogether. I dunno. There's room for improvements with the `vars` handler I think. Let's scrap the `[ ]` idea for now :joy:

Comment (armadi1809) [reply caddytest/integration/caddyfile_adapt/log_filter_with_header.txt]: @francislavoie Sorry for the delay! I didn't have access to a computer in the past week. I took a stab at implementing the new `log_name` directive. It seems to work fine with the vars handler, though. Am I missing something?

File: caddyconfig/httpcaddyfile/builtins.go
Comment (francislavoie) [inline caddyconfig/httpcaddyfile/builtins.go:1075]: I think this should use the `AccessLoggerNameVarKey` const here

File: caddytest/integration/caddyfile_adapt/log_filter_with_header.txt
Comment (francislavoie) [inline caddytest/integration/caddyfile_adapt/log_filter_with_header.txt:15]: Can drop the double-quotes here (doesn't really matter but just saying)

File: caddytest/integration/caddyfile_adapt/log_filter_with_header.txt
Comment (francislavoie) [inline caddytest/integration/caddyfile_adapt/log_filter_with_header.txt:95]: Oh good, I wasn't sure if the `vars` handler would just take an array as-is. That's nice :+1:

File: modules/caddyhttp/logging.go
Comment (francislavoie) [inline modules/caddyhttp/logging.go:84]: Should this type assert as `[]string` instead? Then you don't need to do `.(string)` later, I think.

Comment (francislavoie) [reply modules/caddyhttp/logging.go:84]: I'm wrong, that doesn't work. Go is silly, can't cast directly to a slice type.

**Comment by francislavoie** [comment]
I did a commit to do a final bit of cleanup and refactoring. I think this is good to go now.

**Review by francislavoie — APPROVED** [review]

**Comment by mholt** [comment]
Awesome, thank you both!!

**Comment by steffenbusch** [comment]
Just wanted to say THANK YOU for this awesome new feature. I have just tested it with Caddy v2.8.0-rc and my configuration looked like this (maybe this helps for other users, viewers of this MR or for the caddy docs website).

```caddy
example.com {
        log example.com

        log health_check_log {
                output file /var/log/caddy/access-health-checks.json {
                        roll_size 50m
                        roll_local_time
                        roll_keep_for 30d
                }
                format filter {
                        wrap json
                        fields {
                                request>tls>version tls_version TLSv
                                request>tls>cipher_suite tls_cipher
                        }
                }
                no_hostname
        }

	@monitoring_ohdear_uptime {
		# Two possible user-agent headers:
		# Mozilla/5.0 (compatible; OhDear/1.1; +https://ohdear.app/checker; brokenLinks)
		# OhDear.app (+https://ohdear.app/docs/checks/uptime)
		header_regexp user-agent `^.*\+https:\/\/ohdear\.app.*$`
		header_regexp X-External-Monitoring ^redacted_secret$
	}

        # Before Caddy v.2.8.0 these requests were skipped; now commented
        #log_skip @monitoring_ohdear_uptime

        # Now with Caddy v2.8.0-rc.1 these requests are written into a dedicated file 👍 
        log_name @monitoring_ohdear_uptime health_check_log
 ...
```

**Pull request merged**
